### PR TITLE
KEYCLOAK-14304 Expand xml parsing for SAML Identity Provider

### PR DESCRIPTION
**Pull request by cgeorgilakis**
Changes discribed in https://issues.redhat.com/browse/KEYCLOAK-14304

### CODE REVIEW

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLAssertionParser.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.keycloak.saml.processing.core.parsers.saml.metadata;
+
+public class SAMLAssertionParser extends SAMLAttributeTypeParser {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLAssertionParser.java:3]: Remove and use parser from `org.keycloak.saml.processing.core.parsers.saml.assertion`

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLAttributeParser.java [hunk 2]
```diff
@@ -16,20 +16,12 @@
...
-import javax.xml.stream.XMLEventReader;
-import javax.xml.stream.events.StartElement;
-
 /**
  * Parse the <conditions> in the saml assertion
  *
  * @since Oct 14, 2010
  */
-public class SAMLAttributeParser extends AbstractStaxSamlMetadataParser<AttributeType> {
+public class SAMLAttributeParser extends SAMLAttributeTypeParser {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLAttributeParser.java:24]: Remove and use parser from `org.keycloak.saml.processing.core.parsers.saml.assertion`

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLAttributeTypeParser.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package org.keycloak.saml.processing.core.parsers.saml.metadata;
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLAttributeTypeParser.java:1]: Remove this change

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java [hunk 4]
```diff
@@ -152,6 +155,18 @@ public SAMLIdentityProviderConfig createConfig() {
                             samlIdentityProviderConfig.setEncryptionPublicKey(defaultCertificate);
                         }
                     }
+                    
+                    samlIdentityProviderConfig.setEnabledConfig(entityType.getValidUntil() == null || entityType.getValidUntil().toGregorianCalendar().getTime().after(new Date()) );
+                    
+                	//check for hide on login attibute
+        			if ( entityType.getExtensions() != null && entityType.getExtensions().getEntityAttributes() != null ) {
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java:162]: Check the formatting.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java [hunk 5]
```diff
@@ -52,6 +52,8 @@
     public static final String WANT_ASSERTIONS_SIGNED = "wantAssertionsSigned";
     public static final String WANT_AUTHN_REQUESTS_SIGNED = "wantAuthnRequestsSigned";
     public static final String XML_SIG_KEY_INFO_KEY_NAME_TRANSFORMER = "xmlSigKeyInfoKeyNameTransformer";
+    public static final String HIDE_ON_LOGIN = "hideOnLoginPage";
+    public static final String ENABLED = "enabled";
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java:56]: Remove, there is already `enabled` property in the parent class.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java [hunk 6]
```diff
@@ -52,6 +52,8 @@
     public static final String WANT_ASSERTIONS_SIGNED = "wantAssertionsSigned";
     public static final String WANT_AUTHN_REQUESTS_SIGNED = "wantAuthnRequestsSigned";
     public static final String XML_SIG_KEY_INFO_KEY_NAME_TRANSFORMER = "xmlSigKeyInfoKeyNameTransformer";
+    public static final String HIDE_ON_LOGIN = "hideOnLoginPage";
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java:55]: This constant should be in `IdentityProviderModel`.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLExtensionsParser.java [hunk 7]
```diff
@@ -47,6 +47,14 @@ protected ExtensionsType instantiateElement(XMLEventReader xmlEventReader, Start
 
     @Override
     protected void processSubElement(XMLEventReader xmlEventReader, ExtensionsType target, SAMLMetadataQNames element, StartElement elementDetail) throws ParsingException {
-        target.addExtension(StaxParserUtil.getDOMElement(xmlEventReader));
+        
+    	switch (element) {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLExtensionsParser.java:51]: Check formatting.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLEntityAttributesParser.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import javax.xml.stream.XMLEventReader;
+import javax.xml.stream.events.StartElement;
+
+import org.keycloak.dom.saml.v2.assertion.AttributeType;
+import org.keycloak.dom.saml.v2.metadata.EntityAttributes;
+import org.keycloak.saml.common.exceptions.ParsingException;
+import org.keycloak.saml.common.util.StaxParserUtil;
+
+public class SAMLEntityAttributesParser extends AbstractStaxSamlMetadataParser<EntityAttributes> implements Serializable {
+	 private static final SAMLEntityAttributesParser INSTANCE = new SAMLEntityAttributesParser();
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLEntityAttributesParser.java:16]: Check formatting

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js [hunk 9]
```diff
@@ -965,6 +965,7 @@ module.controller('RealmIdentityProviderCtrl', function($scope, $filter, $upload
         for (var key in data) {
             $scope.identityProvider.config[key] = data[key];
         }
+        $scope.identityProvider.enabled = data["enabled"];
```
Comment (hmlnarik) [inline themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js:968]: This change is not needed.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you for your contribution, this seems like a useful one.

There are few comments to address, please see here and inline.

Most importantly, http://docs.oasis-open.org/security/saml/Post2.0/sstc-metadata-attr-cd-01.pdf states that
> The <mdattr:EntityAttributes> element is a wrapper for one or more <saml:Attribute> or
<saml:Assertion> elements.
Those elements originate in `saml`, i.e. `urn:oasis:names:tc:SAML:2.0:assertion ` namespace. Hence there should be no new SAMLAssertionParser.java nor SAMLAttributeParser.java in the `metadata` package, those from `assertion` package must be reused, please remove them.

Please check formatting of the new files and use spaces, not tabs, with 4 spaces per indentation level, to match the rest of the codebase.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java (hunk 5 above)
Comment (cgeorgilakis) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java:56]: I know it. The problem is that importFrom method in class IdentityProvidersResource ( rest api for import identity provider from url or file) is returning only IdentityProviderModel.config. Property enabled in IdentityProviderModel is not in this map. As a result when you import a Saml Identity Provider from url or file , property enabled will not returned from rest api. Do you suggest another solution?

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js (hunk 9 above)
Comment (cgeorgilakis) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js:968]: Related to my comment in SAMLIdentityProviderConfig, data returned from rest api return only Map. "enabled" key containing in data Map will never affect identityProvider.enabled with previous code. Do you suggest another solution?

My current implementation need a change in order to take into account that "enabled" key will not exist in Map for all other Identity providers except than SAML Identity Provider. I proposed to be : 
`if (data["enabled"] != undefined) {
        	$scope.identityProvider.enabled = data["enabled"];
        }`

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLAttributeParser.java (hunk 2 above)
Comment (cgeorgilakis) [reply saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLAttributeParser.java:24]: org.keycloak.saml.processing.core.parsers.saml.assertion.SAMLAttributeParser  will be used in this PR. This SAMLAttributeParser existed in code and it was not used. Do you want to keep org.keycloak.saml.processing.core.parsers.saml.metadata.SAMLAttributeParser same as Keycloak master or to remove it?

**Comment by cgeorgilakis** [comment]
I made requested changes except variable ENABLED in SAMLIdentityProviderConfig and change in realm.js .
Please see my related comments.

After this, I will do squash and rebase from master.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLMetadataQNames.java [hunk 10]
```diff
@@ -77,14 +77,15 @@
     ATTR_IS_REQUIRED(null, "isRequired"),
     ATTR_NAME(null, "Name"),
     ATTR_NAME_FORMAT(null, "NameFormat"),
-
+    ENTITY_ATTRIBUTES(JBossSAMLURIConstants.METADATA_ENTITY_ATTRIBUTES_NSURI, "EntityAttributes"),
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/parsers/saml/metadata/SAMLMetadataQNames.java:80]: This belongs below to L 88, the namespace is other (see comment below)

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java [hunk 11]
```diff
@@ -152,6 +158,18 @@ public SAMLIdentityProviderConfig createConfig() {
                             samlIdentityProviderConfig.setEncryptionPublicKey(defaultCertificate);
                         }
                     }
+                    
+                    samlIdentityProviderConfig.setEnabledConfig(entityType.getValidUntil() == null || entityType.getValidUntil().toGregorianCalendar().getTime().after(new Date()) );
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java:162]: Time needs to be adjusted to local "Keycloak server" time.
```suggestion
                    samlIdentityProviderConfig.setEnabledConfig(entityType.getValidUntil() == null || entityType.getValidUntil().toGregorianCalendar().getTime().after(new Date(Time.currentTime())) );
```

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java (hunk 5 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java:55]: In that case it is necessary to use a specific name, plain `enable` would be confusing with the property above.
```suggestion
    public static final String ENABLED_FROM_METADATA = "enabledFromMetadata";
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java [hunk 12]
```diff
@@ -646,6 +652,10 @@ public void testSamlImportAndExportMultipleSigningKeys() throws URISyntaxExcepti
         form.addFormData("file", body, MediaType.APPLICATION_XML_TYPE, "saml-idp-metadata-two-signing-certs");
 
         Map<String, String> result = realm.identityProviders().importFrom(form);
+        //firtsly check and remove enabled from result
+        boolean enabled = Boolean.valueOf(result.get("enabled"));
+        result.remove("enabled");
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java:657]: Use constant, not literal: 
```suggestion
        boolean enabled = Boolean.valueOf(result.get(ENABLED_FROM_METADATA));
        result.remove(ENABLED_FROM_METADATA);
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java [hunk 13]
```diff
@@ -608,6 +609,11 @@ public void testSamlImportAndExport() throws URISyntaxException, IOException, Pa
         form.addFormData("file", body, MediaType.APPLICATION_XML_TYPE, "saml-idp-metadata.xml");
 
         Map<String, String> result = realm.identityProviders().importFrom(form);
+       //firtsly check and remove enabled from result
+        boolean enabled = Boolean.valueOf(result.get("enabled"));
+        result.remove("enabled");
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java:614]: Use constant, not literal: 
```suggestion
        boolean enabled = Boolean.valueOf(result.get(ENABLED_FROM_METADATA));
        result.remove(ENABLED_FROM_METADATA);
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java [hunk 14]
```diff
@@ -608,6 +609,11 @@ public void testSamlImportAndExport() throws URISyntaxException, IOException, Pa
         form.addFormData("file", body, MediaType.APPLICATION_XML_TYPE, "saml-idp-metadata.xml");
 
         Map<String, String> result = realm.identityProviders().importFrom(form);
+       //firtsly check and remove enabled from result
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java:612]: For the tests:
Add a test with a an IdP that is not enabled and check that it is indeed not enabled after import.

Why is `enabled` removed from the result? It should remain there to test the functionality.

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js [hunk 15]
```diff
@@ -965,6 +965,9 @@ module.controller('RealmIdentityProviderCtrl', function($scope, $filter, $upload
         for (var key in data) {
             $scope.identityProvider.config[key] = data[key];
         }
+        if (data["enabled"] != undefined) {
+            $scope.identityProvider.enabled = data["enabled"];
```
Comment (hmlnarik) [inline themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js:969]:
```suggestion
        if (data["enabledFromMetadata"] !== undefined) {
            $scope.identityProvider.enabled = data["enabledFromMetadata"];

```
(Please note `!==` and not `!=`)

File: saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java [hunk 16]
```diff
@@ -88,4 +89,14 @@ public void removeExtension(Object extension) {
...
-}
\ No newline at end of file
+
+    public EntityAttributes getEntityAttributes() {
+        for (Object o : this.any) {
+            if (o instanceof EntityAttributes) {
+                return (EntityAttributes) o;
+            }
+        }
+        return null;
```
Comment (hmlnarik) [inline saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java:99]:
```suggestion
        return this.any.stream()
            .filter(EntityAttributes::isInstance)
            .findAny();
```

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java [hunk 17]
```diff
@@ -152,6 +158,18 @@ public SAMLIdentityProviderConfig createConfig() {
...
+                    
+                	//check for hide on login attribute
+        			if ( entityType.getExtensions() != null && entityType.getExtensions().getEntityAttributes() != null ) {
+        			    for (AttributeType attribute :  entityType.getExtensions().getEntityAttributes().getAttribute()) {
+        				    if (MACEDIR_ENTITY_CATEGORY.equals(attribute.getName()) && attribute.getAttributeValue().contains(REFEDS_HIDE_FROM_DISCOVERY) )
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java:167]: Prevent potential NPE
```suggestion
        				    if (MACEDIR_ENTITY_CATEGORY.equals(attribute.getName()) && REFEDS_HIDE_FROM_DISCOVERY.equals(attribute.getAttributeValue()))
```

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thank you for the updates, looks much better.

Still please see https://github.com/keycloak/keycloak/blob/master/docs/building.md#working-with-the-codebase and do ensure that the formatting is compatible with the rest of the code. This namely means: 4 spaces per indentation level; don't use tabs but spaces.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java (hunk 17 above)
Comment (cgeorgilakis) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java:167]: attributeValue is a List in class AttributeType which is instantiated as empty list as you can see.
Moreover it is error to check equality between String and List.

File: saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java (hunk 16 above)
Comment (cgeorgilakis) [reply saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java:99]: I agree. However, for adding this I must change maven.compiler.source of this subproject. I believe this is the best solution. Do you want to do it?
Otherwhise a solution could be:

return (EntityAttributes) this.any.stream().filter(o -> o instanceof EntityAttributes).findAny().orElse(null);

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java (hunk 14 above)
Comment (cgeorgilakis) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java:612]: I will add seperate test method for disabled IdP.
The difference between enabledFromMetadata with others properties is that enabledFromMetadata is not existed in config that saved in database. Moreover, assertSamlConfig is executed again after retrieving IdP from database. I will made appropriate changes to test in order to check appropriate value of enabledFromMetadata  and enable field for enabled and disabled IdP.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java (hunk 13 above)
Comment (cgeorgilakis) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java:614]: I strongly agree with you. However, in class IdentityProviderTest only literals are used and I want class writing to be similar.
If you insist, I could add literal for this or for all config properties.

File: saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java (hunk 16 above)
Comment (hmlnarik) [reply saml-core-api/src/main/java/org/keycloak/dom/saml/v2/metadata/ExtensionsType.java:99]: Ah, ignore it for this moment then, I didn't realize this is still used by JDK 7 code.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java (hunk 17 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java:167]: You're indeed right :-)

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java (hunk 13 above)
Comment (hmlnarik) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/IdentityProviderTest.java:614]: Thank you. Constant just for this single change would set a good example for the rest, but those needn't be addressed in this PR.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java (hunk 11 above)
Comment (cgeorgilakis) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java:162]: new Date() return Keycloak server local Date time.
Time class you mean org.keycloak.common.util.Time? This method return second and Java Date(long date) wait for milliseconds.
If you insist I could write it as :
 samlIdentityProviderConfig.setEnabledConfig(entityType.getValidUntil() == null || entityType.getValidUntil().toGregorianCalendar().getTime().after(new Date(Time.currentTime() * 1000)) );

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js (hunk 15 above)
Comment (cgeorgilakis) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js:969]: Ok.
Moreover, I observed that if I import an enabled SAML IdP , enabled will be off although enabledFromMetadata will be true.
If I change onoffswitch attribute to onoffswitchvalue ( realm-identity-provider-saml.html line 39) , import will work correctly. However, when admin create or edit SAML Idp, enabled is always off. 
The problem is that identityProvider.enabled is initiated as boolean and data return string as value.
How should I write code correctly without changing app.js ( lines 2236-2366)?

**Comment by cgeorgilakis** [comment]
I did requested changes and reply in some comments.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java (hunk 11 above)
Comment (hmlnarik) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderFactory.java]: Yes please, use `currentTimeMillis()` instead.

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js (hunk 15 above)
Comment (hmlnarik) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js]:
```suggestion
            $scope.identityProvider.enabled = data["enabledFromMetadata"] == "true";
```

Comment (cgeorgilakis) [reply themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js]: You are right. However, if must exists for other Identity Providers like OIDC. In these Identity Providers enabledFromMetadata will not exist in data.

**Comment by cgeorgilakis** [comment]
Squash to one commit

File: themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js [hunk 18]
```diff
@@ -965,6 +965,9 @@ module.controller('RealmIdentityProviderCtrl', function($scope, $filter, $upload
         for (var key in data) {
             $scope.identityProvider.config[key] = data[key];
         }
+        if (data["enabledFromMetadata"] !== undefined ) {
+            $scope.identityProvider.enabled = data["enabledFromMetadata"] == "true";
```
Comment (hmlnarik) [inline themes/src/main/resources/theme/base/admin/resources/js/controllers/realm.js:969]: `enabledFromMetadata` should be also removed from `$scope.identityProvider.config`.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Hi @cgeorgilakis, apologies for late re-review, vacation made it impossible to finish this earlier. Could you please rebase?

**Comment by cgeorgilakis** [comment]
I have done requested chenge, rebase and squash.

**Comment by hmlnarik** [comment]
Thank you. I'm running UI tests in https://keycloak-jenkins.com/job/keycloak-admin-console-matrix/195/
I'll report here whether they passed.

**Comment by hmlnarik** [comment]
> Thank you. I'm running UI tests in [https://keycloak-jenkins.com/job/keycloak-admin-console-matrix/196/](https://keycloak-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/keycloak-admin-console-matrix/196/)
> I'll report here whether they passed.

The tests failed due to the following error:
`Caused by: org.openqa.selenium.SessionNotCreatedException: session not created: This version of ChromeDriver only supports Chrome version 85`
This could have been caused by the codebase being outdated. Could you please rebase to the current master?

When at it, please fix the commit message to start with `KEYCLOAK-14304` followed by brief summary, e.g. `KEYCLOAK-14304 Enhance SAML Identity Provider Metadata processing`

**Comment by cgeorgilakis** [comment]
Rebase and message change was done

**Comment by hmlnarik** [comment]
Thank you. I'm running UI tests in https://keycloak-jenkins.com/job/keycloak-admin-console-matrix/199/

**Review by hmlnarik — APPROVED** [review]

**Comment by hmlnarik** [comment]
Thank you for the contribution!

**Pull request merged**